Sming/Arch/Esp32/Components/driver/include/driver/pwm.h

```diff
@@ -7,6 +7,7 @@
  * pwm.h
  *
  ****/
+#include <soc/soc_caps.h>
```

Comment:
Please get rid of chuff, just require one line in this file:

```
#define PWM_CHANNEL_NUM_MAX SOC_LEDC_CHANNEL_NUM
```

Reply:
without the `#ifdef` even? I mean, I agree, `SOC_LEDC_CHANNEL` *should* be defined from idf

Reply:
The entire file should read:

```
/****
 * Sming Framework Project - Open Source framework for high efficiency native ESP8266 development.
 * Created 2015 by Skurydin Alexey
 * http://github.com/SmingHub/Sming
 * All files of the Sming Core are provided under the LGPL v3 license.
 *
 * pwm.h
 *
 ****/

#pragma once

#include <soc/soc_caps.h>

#define PWM_CHANNEL_NUM_MAX SOC_LEDC_CHANNEL_NUM
```

Reply:
Please make sure `#pragma once` is the first line in the file.

Reply:
You can just paste in the above text.

Reply:
dang, hope I got it now.

Reply:
That is very weird: exactly 2 spaces added to the end of *every* line. What editor are you using?

Reply:
that's vs-code, albeit for quick edits, I'll use vim for quick edits.

Reply:
sometimes I feel sorry that I take so much of your time for such simple changes. 
but then I think: we're getting bugs fixed, so it's not for nothing.

Reply:
Simple changes are often the best place to spend time learning about how a particular software project works. I'm always looking to simplify and clarify wherever possible so forgive me if I seem pedantic. Something I always do is run a visual difference on a PR to check that I haven't touched unrelated lines of code, even if it's only coding style.

Reply:
re vs-code and auto-formatter - could that be integrated into the workspace definition? It would be really helpful to have coding style enforced right there.

Reply:
Configuring vscode for clang-format-8 is very much a Sming development-only thing and users may have other coding styles to consider for their projects which we don't want to mess up.

I'll get the guidance at https://sming.readthedocs.io/en/latest/information/develop/clang-tools.html updated, but for vscode, we need three things:

1. Install `clang-format-8`. A bit tricker now as it's no longer generally available in distros. Have to say the standalone builds are a god-send!
2. Install `clang-format` extension in vscode
3. Configure extension with location of clang-format.

I think this is beyond the scope of `make ide-vscode`.